apps/server/src/modules/sharing/api/import-token.uc.ts

```diff
@@ -0,0 +1,226 @@
+					throw new BadRequestException('Cannot copy board without destination course or room reference');
+				}
+				result = await this.copyColumnBoard(user, shareToken.payload.parentId, destinationId, newName);
+				break;
+			case ShareTokenParentType.Room:
+				result = await this.copyRoom(user, shareToken.payload.parentId, newName);
+				break;
+		}
+
+		return result;
+	}
+
+	private async copyCourse(user: User, courseId: string, newName: string): Promise<CopyStatus> {
+		const requiredPermissions = [Permission.COURSE_CREATE];
+
+		this.authorizationService.checkAllPermissions(user, requiredPermissions);
+
+		const copyStatus = await this.courseCopyService.copyCourse({
+			userId: user.id,
+			courseId,
+			newName,
+		});
+
+		return copyStatus;
+	}
+
+	private async copyLesson(user: User, lessonId: string, courseId: string, copyName?: string): Promise<CopyStatus> {
+		const { course } = await this.shareTokenPermissionService.checkCourseWritePermission(
+			user,
+			courseId,
+			Permission.TOPIC_CREATE
+		);
+
+		const copyStatus = await this.lessonCopyService.copyLesson({
+			user,
+			originalLessonId: lessonId,
+			destinationCourse: course,
+			copyName,
+		});
+
+		return copyStatus;
+	}
+
+	private async copyTask(user: User, originalTaskId: string, courseId: string, copyName?: string): Promise<CopyStatus> {
+		const { course } = await this.shareTokenPermissionService.checkCourseWritePermission(
+			user,
+			courseId,
+			Permission.HOMEWORK_CREATE
+		);
+
+		const copyStatus = await this.taskCopyService.copyTask({
+			user,
+			originalTaskId,
+			destinationCourse: course,
+			copyName,
+		});
+
+		return copyStatus;
+	}
+
+	private async copyColumnBoard(
+		user: User,
+		originalColumnBoardId: string,
+		destinationId: EntityId,
+		copyTitle?: string
+	): Promise<CopyStatus> {
+		const originalBoard = await this.columnBoardService.findById(originalColumnBoardId, 0);
+
+		const targetExternalReference: BoardExternalReference = {
+			id: destinationId,
+			type: originalBoard.context.type,
+		};
+
+		await this.checkBoardContextWritePermission(user, targetExternalReference);
+
+		const sourceStorageLocationReference = await this.getStorageLocationReference(originalBoard.context);
+		const targetStorageLocationReference = await this.getStorageLocationReference(targetExternalReference);
+
+		const copyStatus = this.columnBoardService.copyColumnBoard({
+			originalColumnBoardId,
+			targetExternalReference,
+			sourceStorageLocationReference,
+			targetStorageLocationReference,
+			userId: user.id,
+			copyTitle,
+			targetSchoolId: user.school.id,
+		});
+		return copyStatus;
+	}
+
+	private async copyRoom(user: User, roomId: EntityId, copyName?: string): Promise<CopyStatus> {
+		this.authorizationService.checkOneOfPermissions(user, [Permission.ROOM_CREATE]);
+		const room = await this.roomService.getSingleRoom(roomId);
+		await this.shareTokenPermissionService.checkRoomWritePermission(user, room.id, [Permission.ROOM_SHARE]);
```

Comment:
I don't think that the user needs to have write and share permissions on the source room. The check can be removed.